applications/ShallowWaterApplication/custom_strategies/schemes/residualbased_incrementalupdate_wetting_scheme.h

```diff
@@ -0,0 +1,246 @@
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Miguel Maso Sotomayor
+//
+
+#ifndef KRATOS_RESIDUALBASED_INCREMENTALUPDATE_WETTING_SCHEME_H_INCLUDED
+#define KRATOS_RESIDUALBASED_INCREMENTALUPDATE_WETTING_SCHEME_H_INCLUDED
+
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "processes/process.h"
+#include "solving_strategies/schemes/residualbased_incrementalupdate_static_scheme.h"
+
+
+namespace Kratos
+{
+///@addtogroup ShallowWaterApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+*/
+template<class TSparseSpace, class TDenseSpace >
+class ResidualBasedIncrementalUpdateWettingScheme : public ResidualBasedIncrementalUpdateStaticScheme<TSparseSpace, TDenseSpace>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of ResidualBasedIncrementalUpdateWettingScheme
+    KRATOS_CLASS_POINTER_DEFINITION(ResidualBasedIncrementalUpdateWettingScheme);
+
+    typedef ResidualBasedIncrementalUpdateStaticScheme<TSparseSpace, TDenseSpace> BaseType;
+
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+
+    typedef typename BaseType::TSystemMatrixType TSystemMatrixType;
+
+    typedef typename BaseType::TSystemVectorType TSystemVectorType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// @brief Default constructor.
+    ResidualBasedIncrementalUpdateWettingScheme()
+        : BaseType()
+    {}
+
+    /// @brief Constructor with wetting drying model.
+    ResidualBasedIncrementalUpdateWettingScheme(Process::Pointer pWettingModel)
+        : BaseType()
+        , mpWettingModel(pWettingModel)
+    {}
+
+    /// @brief Copy constructor.
+    explicit ResidualBasedIncrementalUpdateWettingScheme(ResidualBasedIncrementalUpdateWettingScheme& rOther)
+        :BaseType(rOther)
+    {
+    }
+
+    /// Destructor.
+    virtual ~ResidualBasedIncrementalUpdateWettingScheme(){}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Function called once at the beginning of each solution step.
+     * @details The basic operations to be carried in there are the following:
+     * - managing variables to be kept constant over the time step (for example time-Scheme constants depending on the actual time step)
+     * @param rModelPart The model part of the problem to solve
+     * @param rA LHS matrix
+     * @param rDx Incremental update of primary variables
+     * @param rb RHS Vector
+     */
+    void InitializeSolutionStep(
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb
+        ) override
+    {
+        if (mpWettingModel != 0) {
```

Comment:
I would check this once in the Check-function
Also I think the newer way is to check against nullptr

Reply:
I think that you can simply do
```suggestion
        if (!mpWettingModel) {
```

Reply:
Thanks @rubenzorrilla , I have updated it.
@philbucher I don't want to check this once in the Check-function because I want to allow the construction without wetting model. There are other formulations to avoid the zero-depth singularity.